File: src/leetcode/task001TwoSum/Solution.java
package leetcode.task001TwoSum;

import java.util.Arrays;

public class Solution {

    public int[] twoSum(int[] nums, int target) {
        int[] sorted = sortArr(nums);
        int left = 0;
        int right = sorted.length - 1;
        for (; left < right; left++) {
            int search = target - sorted[left];
            for (; (sorted[right] >= search) && (left < right); right--) {
                if (search == sorted[right]) {
                    left = getIndex(nums, sorted[left], 0, nums.length);
                    right = getIndex(nums, sorted[right], nums.length - 1, -1);
                    return new int[]{left, right};
                }
            }
        }
        return null;
    }

    private int[] sortArr(int[] nums) {
        int[] sorted = Arrays.copyOf(nums, nums.length);
        Arrays.sort(sorted);
        return sorted;
    }

    private int getIndex(int[] nums, int element, int from, int to) {
        for (int i = from; i != to; i = (from > to) ? i - 1 : i + 1) {
            if (nums[i] == element) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        var solution = new Solution();
        System.out.println(Arrays.toString(solution.twoSum(new int[]{2, 7, 11, 15}, 9)));
        System.out.println(Arrays.toString(solution.twoSum(new int[]{2, 7}, 9)));
        System.out.println(Arrays.toString(solution.twoSum(new int[]{2, 3, 7, 11, 15}, 9)));
        System.out.println(Arrays.toString(solution.twoSum(new int[]{2, 3, 11, 15}, 9)));
        System.out.println(Arrays.toString(solution.twoSum(new int[]{3, 3}, 6)));
        System.out.println(Arrays.toString(solution.twoSum(new int[]{9, 3, 3, 2}, 6)));
    }
}

File: src/main/leetcode/task001TwoSum/Solution.kt
package leetcode.task001TwoSum

import java.util.*

fun twoSum(nums: IntArray, target: Int): IntArray? {
  val sorted = sort(nums)
  var left = 0
  var right = nums.size - 1

  while (left < right) {
    val search = target - sorted[left]
    while (sorted[right] >= search && left < right) {
      if (search == sorted[right]) {
        left = getIndex(nums, sorted[left], 0, nums.size)
        right = getIndex(nums, sorted[right], nums.size - 1, -1)
        return intArrayOf(left, right)
      }
      right--
    }
    left++
  }
  return null
}

val sort: (IntArray) -> IntArray = {
  val sortedArray = it.clone()
  Arrays.sort(sortedArray)
  sortedArray
}

fun getIndex(nums: IntArray, element: Int, from: Int, to: Int): Int {
  var i = from
  while (i != to) {
    if (nums[i] == element) {
      return i
    }
    i = if (from > to) i - 1 else i + 1
  }
  return -1
}

fun main() {
  println(twoSum(intArrayOf(2, 7, 11, 15), 9)?.contentEquals(intArrayOf(0, 1)))
  println(twoSum(intArrayOf(2, 7), 9)?.contentEquals(intArrayOf(0, 1)))
  println(twoSum(intArrayOf(2, 3, 7, 11, 15), 9)?.contentEquals(intArrayOf(0, 2)))
  println(twoSum(intArrayOf(2, 3, 11, 15), 9) == null)
  println(twoSum(intArrayOf(3, 3), 6)?.contentEquals(intArrayOf(0, 1)))
  println(twoSum(intArrayOf(9, 3, 3, 2), 6)?.contentEquals(intArrayOf(1, 2)))
}